File: PlacementCellBackend/Controllers/JobsandInterships/ResearchOpeningsController.cs
using Microsoft.AspNetCore.Mvc;
using PlacementCellBackend.DTOs.ReasearchOpeningsDtos;
using PlacementCellBackend.Services.Placements.Interfaces;

namespace PlacementCellBackend.Controllers.JobsandInterships;

[ApiController]
[Route("api/[controller]")]
public class ResearchOpeningsController : ControllerBase
{
    private readonly ITeacherPlacementService _teacherPlacementService;

    public ResearchOpeningsController(ITeacherPlacementService teacherPlacementService)
    {
        _teacherPlacementService = teacherPlacementService;
    }

    [HttpGet]
    public async Task<ActionResult<IEnumerable<ResearchOpeningDto>>> GetAllTeacherResearchOpenings()
    {
        var openings = await _teacherPlacementService.GetAllTeacherResearchOpeningsAsync();
        return Ok(openings);
    }

    [HttpGet("{id}")]
    public async Task<ActionResult<ResearchOpeningDto>> GetTeacherResearchOpeningById(int id)
    {
        var opening = await _teacherPlacementService.GetTeacherResearchOpeningByIdAsync(id);
        if (opening == null)
            return NotFound();
        return Ok(opening);
    }

    [HttpPost]
    public async Task<ActionResult<ResearchOpeningCreateDto>> CreateTeacherResearchOpening(ResearchOpeningCreateDto researchOpening)
    {
        var created = await _teacherPlacementService.CreateTeacherResearchOpeningAsync(researchOpening);
        return CreatedAtAction(nameof(GetTeacherResearchOpeningById), new { }, created);
    }

    [HttpPut("{id}")]
    public async Task<IActionResult> UpdateTeacherResearchOpening(int id, ResearchOpeningCreateDto updatedResearchOpening)
    {

        var success = await _teacherPlacementService.UpdateTeacherResearchOpeningAsync(id, updatedResearchOpening);
        if (!success)
            return NotFound();

        return NoContent();
    }

    [HttpDelete("{id}")]
    public async Task<IActionResult> DeleteTeacherResearchOpening(int id)
    {
        var success = await _teacherPlacementService.DeleteTeacherResearchOpeningAsync(id);
        if (!success)
            return NotFound();

        return NoContent();
    }
}

File: PlacementicsBackend/src/main/java/com/example/placementicsbackend/controllers/JobsandInterships/ResearchOpeningsController.java
package com.example.placementicsbackend.controllers.JobsandInterships;

import com.example.placementicsbackend.dtos.ReasearchOpeningsDtos.ResearchOpeningCreateDto;
import com.example.placementicsbackend.dtos.ReasearchOpeningsDtos.ResearchOpeningDto;
import com.example.placementicsbackend.services.placements.interfaces.ITeacherPlacementService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/research-openings")
@RequiredArgsConstructor
public class ResearchOpeningsController {

    private final ITeacherPlacementService teacherPlacementService;

    @GetMapping
    public ResponseEntity<List<ResearchOpeningDto>> getAllTeacherResearchOpenings() {
        List<ResearchOpeningDto> openings =
                teacherPlacementService.getAllTeacherResearchOpenings();

        return ResponseEntity.ok(openings);
    }

    @GetMapping("/{id}")
    public ResponseEntity<ResearchOpeningDto> getTeacherResearchOpeningById(
            @PathVariable Integer id) {

        ResearchOpeningDto opening =
                teacherPlacementService.getTeacherResearchOpeningById(id);

        if (opening == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(opening);
    }

    @PostMapping
    public ResponseEntity<ResearchOpeningDto> createTeacherResearchOpening(
            @RequestBody ResearchOpeningCreateDto researchOpening) {

        ResearchOpeningDto created =
                teacherPlacementService.createTeacherResearchOpening(researchOpening);

        return ResponseEntity
                .status(HttpStatus.CREATED)
                .body(created);
    }

    @PutMapping("/{id}")
    public ResponseEntity<Void> updateTeacherResearchOpening(
            @PathVariable Integer id,
            @RequestBody ResearchOpeningCreateDto updatedResearchOpening) {

        boolean success =
                teacherPlacementService.updateTeacherResearchOpening(id, updatedResearchOpening);

        if (!success) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.noContent().build();
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteTeacherResearchOpening(
            @PathVariable Integer id) {

        boolean success =
                teacherPlacementService.deleteTeacherResearchOpening(id);

        if (!success) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.noContent().build();
    }
}